File: backend/app/api/v1/prices.py
from typing import Optional
from fastapi import APIRouter, Depends, Query, status
from app.core.deps import require_admin
from app.schemas.price import (
    PriceCreateRequest,
    PriceStatusRequest,
    PriceUpdateRequest,
)
from app.services.price_service import price_service
from app.utils.responses import success_response

router = APIRouter(tags=["Master Harga Sampah"])


@router.get("/master/waste-prices")
def list_prices(
    category_id: Optional[str] = Query(None, description="Filter per kategori"),
    status: Optional[str] = Query(None, description="Filter status ACTIVE/INACTIVE"),
    search: Optional[str] = Query(None, description="Pencarian kategori atau catatan"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100)
):
    """Daftar harga sampah dengan filter kategori, status, dan pagination."""
    items, total_items = price_service.list_prices(
        category_id=category_id,
        status_filter=status,
        search=search,
        page=page,
        page_size=page_size
    )
    total_pages = (total_items + page_size - 1) // page_size if total_items > 0 else 0
    return success_response(
        data={
            "items": items,
            "pagination": {
                "page": page,
                "page_size": page_size,
                "total_items": total_items,
                "total_pages": total_pages,
            }
        }
    )


@router.get("/master/categories/{category_id}/active-price")
def get_category_active_price(category_id: str):
    """Mendapatkan harga aktif terkini untuk kategori sampah tertentu."""
    price = price_service.get_active_price_by_category(category_id)
    return success_response(data=price)


@router.post("/admin/master/waste-prices", status_code=status.HTTP_201_CREATED)
def create_price(data: PriceCreateRequest, admin_user: dict = Depends(require_admin)):
    """Menetapkan harga sampah baru (Admin). Jika status ACTIVE, harga aktif lama otomatis nonaktif."""
    price = price_service.create_price(data, creator_id=admin_user["id"])
    return success_response(
        data=price,
        message="Master harga sampah berhasil disimpan.",
        status_code=status.HTTP_201_CREATED
    )


@router.put("/admin/master/waste-prices/{price_id}")
def update_price(
    price_id: str, data: PriceUpdateRequest, admin_user: dict = Depends(require_admin)
):
    """Ubah data harga sampah (Admin)."""
    price = price_service.update_price(price_id, data, updater_id=admin_user["id"])
    return success_response(data=price, message="Master harga sampah berhasil diperbarui.")


@router.patch("/admin/master/waste-prices/{price_id}/status")
def update_price_status(
    price_id: str, data: PriceStatusRequest, admin_user: dict = Depends(require_admin)
):
    """Ubah status ACTIVE/INACTIVE harga sampah (Admin)."""
    price = price_service.update_price_status(price_id, data.status, updater_id=admin_user["id"])
    return success_response(
        data=price,
        message=f"Status harga berhasil diubah menjadi {data.status}."
    )


@router.delete("/admin/master/waste-prices/{price_id}")
def delete_price(price_id: str, admin_user: dict = Depends(require_admin)):
    """Hapus master harga sampah secara permanen (Admin)."""
    price_service.delete_price(price_id)
    return success_response(message="Master harga sampah berhasil dihapus.")

File: backend/src/routes/prices.ts
import { Hono } from 'hono';
import { AppEnv } from '../types/index.js';
import { priceService } from '../services/priceService.js';
import { authMiddleware } from '../middleware/auth.js';
import { requireAdmin } from '../middleware/roleGuard.js';
import { successResponse } from '../utils/response.js';
import { validateJson } from '../utils/validator.js';
import {
  priceCreateSchema,
  priceStatusSchema,
  priceUpdateSchema,
} from '../schemas/price.schema.js';

// Public / Authenticated price router mounted at /api/v1/master/waste-prices
export const pricesRouter = new Hono<AppEnv>();

pricesRouter.get('/', async (c) => {
  const query = c.req.query();
  const result = await priceService.listPrices({
    category_id: query.category_id,
    status: query.status,
    search: query.search,
    page: query.page ? parseInt(query.page, 10) : 1,
    page_size: query.page_size ? parseInt(query.page_size, 10) : 20,
  });
  return successResponse(c, result, 'Master harga berhasil dimuat.');
});

pricesRouter.get('/:id', async (c) => {
  const id = c.req.param('id');
  const result = await priceService.getPriceById(id);
  return successResponse(c, result, 'Detail harga berhasil dimuat.');
});

// Admin price router mounted at /api/v1/admin/master/waste-prices
export const adminPricesRouter = new Hono<AppEnv>();
adminPricesRouter.use('*', authMiddleware, requireAdmin);

adminPricesRouter.post('/', validateJson(priceCreateSchema), async (c) => {
  const user = c.get('user');
  const data = c.req.valid('json' as any);
  const result = await priceService.createPrice(data, user.id);
  return successResponse(c, result, 'Master harga berhasil ditetapkan.', 201);
});

adminPricesRouter.put('/:id', validateJson(priceUpdateSchema), async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const data = c.req.valid('json' as any);
  const result = await priceService.updatePrice(id, data, user.id);
  return successResponse(c, result, 'Master harga berhasil diperbarui.');
});

adminPricesRouter.patch('/:id/status', validateJson(priceStatusSchema), async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const data = c.req.valid('json' as any);
  const result = await priceService.updatePriceStatus(id, data.status, user.id);
  return successResponse(c, result, 'Status harga berhasil diubah.');
});

adminPricesRouter.delete('/:id', async (c) => {
  const id = c.req.param('id');
  await priceService.deletePrice(id);
  return successResponse(c, null, 'Master harga berhasil dihapus.');
});
